File: Depth-Perception-Using-Stereo-Camera/python/tests/test_stereo_depth.py
from __future__ import annotations

import subprocess
import sys
from dataclasses import replace
from pathlib import Path

import cv2
import numpy as np
import pytest


PYTHON_DIR = Path(__file__).resolve().parents[1]
PROJECT_DIR = PYTHON_DIR.parent
sys.path.insert(0, str(PYTHON_DIR))

from stereo_depth import (  # noqa: E402
    RectificationMaps,
    StereoBMConfig,
    compute_disparity,
    disparity_to_depth,
    find_largest_obstacle,
    fit_depth_model,
    load_config,
    load_rectification_maps,
    save_config,
)


def synthetic_pair(disparity: int = 8) -> tuple[np.ndarray, np.ndarray]:
    random = np.random.default_rng(7)
    left = random.integers(0, 256, (120, 192), dtype=np.uint8)
    right = np.zeros_like(left)
    right[:, :-disparity] = left[:, disparity:]
    right[:, -disparity:] = random.integers(
        0, 256, (left.shape[0], disparity), dtype=np.uint8
    )
    return left, right


def synthetic_config() -> StereoBMConfig:
    return StereoBMConfig(
        num_disparities=16,
        block_size=9,
        pre_filter_type=1,
        pre_filter_size=9,
        pre_filter_cap=31,
        texture_threshold=0,
        uniqueness_ratio=0,
        speckle_range=0,
        speckle_window_size=0,
        disp12_max_diff=-1,
        min_disparity=0,
        depth_scale=1000.0,
        depth_offset=5.0,
    )


def test_legacy_config_is_validated_and_migrated() -> None:
    path = PROJECT_DIR / "data" / "depth_estmation_params_py.xml"
    config = load_config(path)
    assert config.num_disparities == 144
    assert config.block_size == 35
    assert config.depth_scale == pytest.approx(36.51238116534359 * 144)
    assert config.depth_offset == 0.0


def test_config_round_trip_preserves_scale_and_offset(tmp_path: Path) -> None:
    expected = synthetic_config()
    path = save_config(expected, tmp_path / "config.xml")
    assert load_config(path) == expected


@pytest.mark.parametrize(
    "config",
    [
        replace(synthetic_config(), num_disparities=15),
        replace(synthetic_config(), block_size=8),
        replace(synthetic_config(), pre_filter_cap=0),
        replace(synthetic_config(), depth_scale=float("nan")),
    ],
)
def test_invalid_config_is_rejected(config: StereoBMConfig) -> None:
    with pytest.raises(ValueError):
        config.validated()


def test_rectification_maps_are_complete() -> None:
    maps = load_rectification_maps()
    assert maps.image_size == (640, 480)
    assert maps.left_x.shape[:2] == (480, 640)
    assert maps.right_x.shape[:2] == (480, 640)


def test_invalid_rectification_map_type_is_rejected() -> None:
    invalid = np.zeros((10, 10), dtype=np.float64)
    maps = RectificationMaps(invalid, invalid, invalid, invalid)
    with pytest.raises(ValueError, match="Rectification maps"):
        maps.validated()


def test_synthetic_stereobm_recovers_known_shift() -> None:
    left, right = synthetic_pair(disparity=8)
    disparity = compute_disparity(left, right, synthetic_config())
    assert np.mean(np.isfinite(disparity)) > 0.75
    assert float(np.nanmedian(disparity)) == pytest.approx(8.0, abs=0.25)


def test_depth_model_fit_preserves_slope_and_intercept() -> None:
    disparities = np.array([8.0, 10.0, 16.0, 25.0, 40.0])
    depths = 1200.0 / (disparities - 2.0) + 17.0
    scale, offset, rmse = fit_depth_model(
        disparities, depths, min_disparity=2.0
    )
    assert scale == pytest.approx(1200.0)
    assert offset == pytest.approx(17.0)
    assert rmse < 1e-10


def test_invalid_disparity_becomes_nan_depth() -> None:
    config = replace(synthetic_config(), min_disparity=2)
    disparity = np.array([[np.nan, -1.0, 2.0, 7.0]], dtype=np.float32)
    depth = disparity_to_depth(disparity, config)
    assert np.all(np.isnan(depth[0, :3]))
    assert depth[0, 3] == pytest.approx(205.0)


def test_obstacle_detection_is_deterministic() -> None:
    depth = np.full((100, 100), np.nan, dtype=np.float32)
    depth[20:70, 30:80] = 50.0
    depth[30:40, 40:50] = np.nan
    mask, obstacle = find_largest_obstacle(
        depth,
        min_depth=10.0,
        max_depth=100.0,
        minimum_area_fraction=0.1,
    )
    assert obstacle is not None
    assert obstacle.bounding_box == (30, 20, 50, 50)
    assert obstacle.mean_depth == pytest.approx(50.0)
    assert np.count_nonzero(mask) == 2400


def test_headless_cli_saves_outputs(tmp_path: Path) -> None:
    left, right = synthetic_pair()
    left_path = tmp_path / "left.png"
    right_path = tmp_path / "right.png"
    assert cv2.imwrite(str(left_path), left)
    assert cv2.imwrite(str(right_path), right)
    config_path = save_config(synthetic_config(), tmp_path / "config.xml")
    output_dir = tmp_path / "results"

    result = subprocess.run(
        [
            sys.executable,
            str(PYTHON_DIR / "stereo_depth.py"),
            "--left",
            str(left_path),
            "--right",
            str(right_path),
            "--config",
            str(config_path),
            "--already-rectified",
            "--output-dir",
            str(output_dir),
        ],
        check=False,
        capture_output=True,
        text=True,
    )
    assert result.returncode == 0, result.stderr
    for name in (
        "left-rectified.png",
        "right-rectified.png",
        "disparity.png",
        "depth.png",
    ):
        assert (output_dir / name).is_file()

File: Depth-Perception-Using-Stereo-Camera/cpp/tests/test_stereo_depth.cpp
#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <vector>

#include <opencv2/core.hpp>

#include "stereo_depth.hpp"

namespace {

void require(bool condition, const char* message) {
    if (!condition) {
        throw std::runtime_error(message);
    }
}

void requireNear(
    double actual, double expected, double tolerance,
    const char* message) {
    if (std::abs(actual - expected) > tolerance) {
        throw std::runtime_error(message);
    }
}

learnopencv::stereo::StereoBMConfig syntheticConfig() {
    learnopencv::stereo::StereoBMConfig config;
    config.numDisparities = 16;
    config.blockSize = 9;
    config.preFilterType = 1;
    config.preFilterSize = 9;
    config.preFilterCap = 31;
    config.textureThreshold = 0;
    config.uniquenessRatio = 0;
    config.speckleRange = 0;
    config.speckleWindowSize = 0;
    config.disp12MaxDiff = -1;
    config.minDisparity = 0;
    config.depthScale = 1000.0;
    config.depthOffset = 5.0;
    config.validate();
    return config;
}

std::pair<cv::Mat, cv::Mat> syntheticPair(int disparity = 8) {
    cv::RNG random(7);
    cv::Mat left(120, 192, CV_8UC1);
    random.fill(left, cv::RNG::UNIFORM, 0, 256);
    cv::Mat right = cv::Mat::zeros(left.size(), left.type());
    left.colRange(disparity, left.cols).copyTo(
        right.colRange(0, right.cols - disparity));
    random.fill(
        right.colRange(right.cols - disparity, right.cols),
        cv::RNG::UNIFORM, 0, 256);
    return {left, right};
}

}  // namespace

int main() {
    try {
        const auto source =
            std::filesystem::path(STEREO_SOURCE_DIR);
        const auto legacy = learnopencv::stereo::loadConfig(
            source / "data" / "depth_estimation_params_cpp.xml");
        require(legacy.numDisparities == 144,
                "Legacy numDisparities mismatch.");
        requireNear(
            legacy.depthScale, 41.32084274291992 * 144.0, 1e-8,
            "Legacy depth scale migration failed.");

        const auto maps =
            learnopencv::stereo::loadRectificationMaps(
                source / "data" / "stereo_rectify_maps.xml");
        require(
            maps.imageSize() == cv::Size(640, 480),
            "Rectification map size mismatch.");
        learnopencv::stereo::RectificationMaps invalidMaps;
        invalidMaps.leftX = cv::Mat::zeros(10, 10, CV_8UC1);
        invalidMaps.leftY = cv::Mat::zeros(10, 10, CV_8UC1);
        invalidMaps.rightX = cv::Mat::zeros(10, 10, CV_8UC1);
        invalidMaps.rightY = cv::Mat::zeros(10, 10, CV_8UC1);
        bool invalidMapsRejected = false;
        try {
            invalidMaps.validate();
        } catch (const std::invalid_argument&) {
            invalidMapsRejected = true;
        }
        require(
            invalidMapsRejected,
            "Invalid rectification map types were accepted.");

        auto invalid = syntheticConfig();
        invalid.numDisparities = 15;
        bool invalidRejected = false;
        try {
            invalid.validate();
        } catch (const std::invalid_argument&) {
            invalidRejected = true;
        }
        require(invalidRejected, "Invalid StereoBM config was accepted.");

        const auto config = syntheticConfig();
        const auto temporary =
            std::filesystem::current_path() /
            "stereo-depth-roundtrip.xml";
        learnopencv::stereo::saveConfig(config, temporary);
        const auto roundTrip =
            learnopencv::stereo::loadConfig(temporary);
        std::filesystem::remove(temporary);
        requireNear(
            roundTrip.depthScale, config.depthScale, 1e-12,
            "Config scale round trip failed.");
        requireNear(
            roundTrip.depthOffset, config.depthOffset, 1e-12,
            "Config offset round trip failed.");

        const auto [left, right] = syntheticPair();
        const cv::Mat disparity =
            learnopencv::stereo::computeDisparity(
                left, right, config);
        std::vector<float> finite;
        for (int row = 0; row < disparity.rows; ++row) {
            const auto* values = disparity.ptr<float>(row);
            for (int col = 0; col < disparity.cols; ++col) {
                if (std::isfinite(values[col])) {
                    finite.push_back(values[col]);
                }
            }
        }
        require(
            finite.size() >
                static_cast<std::size_t>(disparity.total() * 0.75),
            "Too few valid synthetic disparities.");
        std::sort(finite.begin(), finite.end());
        requireNear(
            finite[finite.size() / 2], 8.0, 0.25,
            "Synthetic disparity median mismatch.");

        const std::vector<double> disparities{
            8.0, 10.0, 16.0, 25.0, 40.0};
        std::vector<double> depths;
        for (const double value : disparities) {
            depths.push_back(1200.0 / (value - 2.0) + 17.0);
        }
        const auto fit = learnopencv::stereo::fitDepthModel(
            disparities, depths, 2.0);
        requireNear(fit.scale, 1200.0, 1e-8,
                    "Depth fit scale mismatch.");
        requireNear(fit.offset, 17.0, 1e-8,
                    "Depth fit offset mismatch.");
        require(fit.rmse < 1e-10, "Depth fit RMSE is too high.");

        auto depthConfig = config;
        depthConfig.minDisparity = 2;
        cv::Mat sample(
            1, 4, CV_32FC1,
            cv::Scalar(std::numeric_limits<float>::quiet_NaN()));
        sample.at<float>(0, 1) = -1.0F;
        sample.at<float>(0, 2) = 2.0F;
        sample.at<float>(0, 3) = 7.0F;
        const cv::Mat depth =
            learnopencv::stereo::disparityToDepth(
                sample, depthConfig);
        require(std::isnan(depth.at<float>(0, 0)),
                "NaN disparity must remain invalid.");
        require(std::isnan(depth.at<float>(0, 1)),
                "Negative disparity must be invalid.");
        require(std::isnan(depth.at<float>(0, 2)),
                "Minimum disparity must be invalid.");
        requireNear(depth.at<float>(0, 3), 205.0, 1e-5,
                    "Depth conversion mismatch.");

        cv::Mat obstacleDepth(
            100, 100, CV_32FC1,
            cv::Scalar(std::numeric_limits<float>::quiet_NaN()));
        obstacleDepth(
            cv::Rect(30, 20, 50, 50)).setTo(50.0F);
        obstacleDepth(
            cv::Rect(40, 30, 10, 10)).setTo(
                std::numeric_limits<float>::quiet_NaN());
        const auto [mask, obstacle] =
            learnopencv::stereo::findLargestObstacle(
                obstacleDepth, 10.0, 100.0, 0.1);
        require(obstacle.has_value(), "Expected obstacle was not found.");
        require(
            obstacle->boundingBox == cv::Rect(30, 20, 50, 50),
            "Obstacle bounding box mismatch.");
        requireNear(obstacle->meanDepth, 50.0, 1e-6,
                    "Obstacle mean depth mismatch.");
        require(cv::countNonZero(mask) == 2400,
                "Obstacle mask mismatch.");

        std::cout << "All stereo-depth C++ regression checks passed.\n";
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Stereo-depth regression failure: "
                  << error.what() << '\n';
        return 1;
    }
}
